### PR TITLE
Fix double authentication when calling Security_Authenticate with SoapHeader 4

### PR DESCRIPTION
**Pull request by therealartz**
I know that is not urgent at all, but I had some free time and desire to fix it. #151 

Personally I am a little confused with `if` statement that will check message name on every call, but maybe I'm too worried.

In any case, maybe @DerMika  have any idea of how this could be solved in other way?

### CODE REVIEW

File: src/Amadeus/Client/Struct/Security/PasswordInfo.php [hunk 1]
```diff
@@ -43,9 +44,22 @@ class PasswordInfo
...
      */
     const DATA_TYPE_BINARY = 'B';
+    /**
+     * When using "EDIFACT DATA" type, password length value must be 12.
+     *
+     * @var int
+     */
+    const PASSWORD_LENGTH_EDIFACT = 12;
+    /**
+     * When using "BINARY DATA" type, password length value must be 40.
```
Comment (DerMika) [inline src/Amadeus/Client/Struct/Security/PasswordInfo.php]: Are you sure about this? AFAIK this is just the length of the plaintext password, but I don't know if Amadeus uses fixed length passwords (which would not be very smart)

File: src/Amadeus/Client/Session/Handler/SoapHeader4.php [hunk 2]
```diff
@@ -285,81 +285,84 @@ protected function createSoapHeaders($sessionData, $params, $messageName, $messa
...
-            $digest = $this->generatePasswordDigest($password, $creationString, $messageNonce);
-
-            $securityHeaderXml = $this->generateSecurityHeaderRawXml(
-                $params->authParams->userId,
-                $encodedNonce,
-                $digest,
-                $creationString
-            );
+        // If not Security_Authenticate message - send authentication information in headers
+        if ($this->isNotSecurityAuthenticateMessage($messageName)) {
```
Comment (DerMika) [inline src/Amadeus/Client/Session/Handler/SoapHeader4.php]: I would prefer to have only one if stement (combine with the `$this->isAuthenticated === false` below).

**Comment by DerMika** [comment]
I wouldn't worry about the extra if statement too much.

File: src/Amadeus/Client/Struct/Security/PasswordInfo.php (hunk 1 above)
Comment (therealartz) [reply src/Amadeus/Client/Struct/Security/PasswordInfo.php]: For me it was a big surprise.

I used `strlen` and request was:
```
<ns1:passwordInfo>
    <ns1:dataLength>16</ns1:dataLength>
    <ns1:dataType>E</ns1:dataType>
    <ns1:binaryData>UFlOU2ZYV3F6MKdR</ns1:binaryData>
</ns1:passwordInfo>
```
But response was:
```
<freeText>Bad Grammar:  &lt;From: InboundEdifactMessageImpl.cpp#248&gt; Exception caug</freeText>
<freeText>ht while decoding Fusion DSC found an exception !  Binary BLB segment </freeText>
<freeText>  The following structure is expected for BLB:   BLB+binaryLength+data</freeText>
<freeText>Type+binaryData&apos; or its equivalent in xml   with the following restric</freeText>
<freeText>tions :   binaryLength : numeric, must be the exact size in bytes of b</freeText>
<freeText>inaryData   dataType: optional, length 1   binaryData: not checked  Cu</freeText>
<freeText>rrent position in buffer: +PYNSfXWqz5GQ&apos;  Last 15 characters before er</freeText>
<freeText>ror: DUT:SU&apos;BLB+16+E  Current path in the grammar : /passwordInfo#1</freeText>
```

I checked everything several times and also tried with SHA1 hashing of password (which length was bigger than 40).

Amadeus Support replied to me that data length value should be 12.
After I set this fixed data length - authenticate is successful.

Comment (DerMika) [reply src/Amadeus/Client/Struct/Security/PasswordInfo.php]: Whaaaa, okay, thanks for that

**Comment by DerMika** [comment]
Thanks again!

File: src/Amadeus/Client/Struct/Security/PasswordInfo.php (hunk 1 above)
Comment (therealartz) [reply src/Amadeus/Client/Struct/Security/PasswordInfo.php]: Yeah, it's stange

image

Comment (ammuller) [reply src/Amadeus/Client/Struct/Security/PasswordInfo.php:58]: For type E the password length should be the length of the password before it is base64 encoded

**Pull request merged**